src/aleph/http/http2.clj

```diff
@@ -393,18 +699,704 @@
+                             (.fireChannelRead ctx msg))
+                           (swap! frame-list conj msg)))
+
+                       Http2HeadersFrame
+                       (if (.isEndStream ^Http2HeadersFrame msg)
+                         (do
+                           (run! #(.fireChannelRead ctx %) @frame-list)
+                           (.fireChannelRead ctx msg))
+                         (swap! frame-list conj msg))
+
+                       (.fireChannelRead ctx msg))))))
+
+(defn setup-stream-pipeline
+  "Set up the pipeline for an HTTP/2 stream channel"
+  [{:keys [^ChannelPipeline pipeline
+           handler
+           is-server?
+           proxy-options
+           logger
+           http2-stream-pipeline-transform
+           max-request-body-size
+           compression?
+           compression-options]}]
+  (log/trace "setup-stream-pipeline called")
+
+  (when (some? proxy-options)
+    (throw (IllegalArgumentException. "Proxying HTTP/2 messages not supported yet")))
+
+  (cond-> pipeline
+
+          ;; necessary for multipart support in HTTP/2?
+          #_true
+          #_(.addLast "stream-frame-to-http-object"
+                      ^ChannelHandler stream-frame->http-object-codec)
+          true
+          (.addLast "streamer"
+                    ^ChannelHandler (ChunkedWriteHandler.))
+
+          (and is-server? max-request-body-size)
+          (.addLast "max-request-body-size"
+                    (max-size-handler max-request-body-size))
+
+          ;; TODO: add continue handler for server-side
+
+          compression?
+          (.addLast "compression-header-handler"
+                    (compress/h2-compression-handler compression-options))
+
+          true
+          (.addLast "handler"
+                    ^ChannelHandler
+                    handler)
+
+          true
+          (common/add-non-http-handlers logger http2-stream-pipeline-transform)
+
+          #_true
+          #_(common/add-exception-handler "stream-ex-handler"))
+
+  (log/trace "Added all stream handlers")
+  (log/debug "Stream chan pipeline:" pipeline)
+
+  pipeline)
+
+(def ^:private client-inbound-handler
+  "For the client, the inbound handler will probably never get used.
+   Servers will rarely initiate streams, because PUSH_PROMISE is effectively
+   deprecated. Responses to client-initiated streams are set elsewhere
+   (even if it's the same handler fn)."
+  (netty/channel-inbound-handler
+    {:channel-active
+     ([_ ctx]
+      (log/error "Cannot currently handle peer-initiated streams. (You must override this handler if receiving server-pushed streams.) Closing channel.")
+      (netty/close ctx))}))
+
+
+(defn setup-conn-pipeline
+  "Set up pipeline for an HTTP/2 connection channel. Works for both server
+   and client."
+  [{:keys
+    [^ChannelPipeline pipeline
+     ^LoggingHandler logger
+     idle-timeout
+     ^ChannelHandler handler
+     is-server?
```

Comment:
Can we use `server?` across the codebase instead?

Reply:
Good point. That would fit better with the rest of the code.

Alternatively, we could do it Mario-style and rename the var to `its-a-me-server-o`. 😄